src/crypto/tls/handshake_client_tls13.go

```diff
@@ -16,6 +16,44 @@ import (
 	"time"
 )
 
+// EXP_EventTLS13ClientHandshakeTimingInfo carries intra-stack durations for
+// state machine changes. It can be used for tracking metrics during a
+// connection.
+//
+// NOTE: This API is EXPERIMENTAL and subject to change.
+type EXP_EventTLS13ClientHandshakeTimingInfo struct {
+	timer                   func() time.Time
+	start                   time.Time
+	WriteClientHello        time.Duration
+	ReadServerHello         time.Duration
+	ReadEncryptedExtensions time.Duration
+	ReadCertificate         time.Duration
+	ReadCertificateVerify   time.Duration
+	ReadServerFinished      time.Duration
+	WriteClientFinished     time.Duration
+}
+
+// Name is required by the EXP_Event interface.
+func (e EXP_EventTLS13ClientHandshakeTimingInfo) Name() string {
+	return "TLS13ClientHandshakeTimingInfo"
+}
+
+func (e EXP_EventTLS13ClientHandshakeTimingInfo) elapsedTime() time.Duration {
+	return time.Now().Sub(e.start)
+}
+
+func createTLS13ClientHandshakeTimingInfo(timerFunc func() time.Time) EXP_EventTLS13ClientHandshakeTimingInfo {
+	timer := time.Now
+	if timerFunc != nil {
```

Comment:
`timerFunc` is not equal to `nil` for each call to this function. Since this is an internal API, I would just assume this to always be the case and not do this check here.

Reply:
Untrue! Sometimes it is equal to `nil`, depending on what's in the `Config` struct.

Reply:
Oh, good catch.